torch/distributed/device_mesh.py

```diff
@@ -885,66 +688,44 @@ def from_group(
             """
+            if mesh is not None and not isinstance(mesh, torch.Tensor):
+                mesh = torch.tensor(mesh, device="cpu", dtype=torch.int)
 
-            # 1D scenario
             if isinstance(group, ProcessGroup):
                 group_ranks = get_process_group_ranks(group)
-                if (
-                    isinstance(mesh, torch.Tensor) and mesh.tolist() != group_ranks
-                ) or (
-                    mesh is not None
-                    and not isinstance(mesh, torch.Tensor)
-                    and mesh != group_ranks
-                ):
+                if mesh is None:
+                    mesh = torch.tensor(group_ranks, device="cpu", dtype=torch.int)
+                elif mesh.tolist() != group_ranks:
                     raise ValueError(
-                        f"Invalid mesh {str(mesh)} for ProcessGroup with ranks {group_ranks}"
+                        f"Invalid mesh {mesh.tolist()} for ProcessGroup with ranks {group_ranks}"
                     )
-                mesh = torch.tensor(group_ranks, device="cpu", dtype=torch.int)
-                device_mesh = DeviceMesh(
-                    device_type,
-                    mesh,
-                    mesh_dim_names=mesh_dim_names,
-                    _init_backend=False,
-                )
-                device_mesh._dim_group_names = [group.group_name]
-                return device_mesh
-
-            # nD scenario
-            groups = list(group)
-            if len(groups) == 0:
-                raise ValueError("Expects at least one ProcessGroup to be passed")
-            if mesh is None:
-                raise ValueError("Must pass mesh if passing multiple ProcessGroups")
-            if mesh_dim_names is None:
-                raise ValueError(
-                    "Must pass mesh_dim_names if passing multiple ProcessGroups"
-                )
-            mesh = (
-                mesh.detach().to(dtype=torch.int, device="cpu")
-                if isinstance(mesh, torch.Tensor)
-                else torch.tensor(mesh, device="cpu", dtype=torch.int)
-            )
-            if mesh.ndim != len(groups):
+                group = [group]
+
+            if len(group) != mesh.ndim:
                 raise ValueError(
-                    "Expects mesh with ndim equal to number of ProcessGroups but got "
-                    f"mesh {mesh.tolist()} and {len(groups)} ProcessGroups"
+                    f"Expects mesh with ndim equal to number of ProcessGroups but got mesh {mesh.tolist()} and {len(group)} ProcessGroups"
                 )
-            device_mesh = DeviceMesh(
-                device_type, mesh, mesh_dim_names=mesh_dim_names, _init_backend=False
-            )
-            device_mesh._dim_group_names = [group.group_name for group in groups]
-            return device_mesh
+
+            storage = DeviceMeshStorage(device_type, _init_backend=False)
```

Comment:
Oh so the idea is that there's no explicit "singleton" storage, but rather explicitly now each mesh tensor <-> storage. I wonder would the same `mesh_tensor` result in the same DeviceMeshStorage?

Reply:
I see, I think the contract make sense! Then yeah we can just make the test to assert that created from `from_group` be different from the one directly created by `DeviceMesh`